sdk/python/approzium/psycopg2.py

```diff
@@ -0,0 +1,107 @@
+import psycopg2
+import select
+import socket
+import struct
+import logging
+from ctypes import cdll, create_string_buffer
+from ctypes.util import find_library
+from .socketfromfd import fromfd
+from .authenticator import get_hash
+from .misc import read_int32_from_bytes
+
+libpq = cdll.LoadLibrary("libpq.so.5")
+libssl = cdll.LoadLibrary(find_library("ssl"))
+
+
+def advance_connection(pgconn):
+    state = pgconn.poll()
+    status = libpq.PQstatus(pgconn.pgconn_ptr)
+    fd = pgconn.fileno()
+    if state == psycopg2.extensions.POLL_OK:
+        pass
+    elif state == psycopg2.extensions.POLL_WRITE:
+        select.select([], [fd], [])
+    elif state == psycopg2.extensions.POLL_READ:
+        select.select([fd], [], [])
+    else:
+        raise psycopg2.OperationalError("poll() returned %s" % state)
+    return state, status
+
+
+def advance_until_challenge(pgconn):
+    CONNECTION_AWAITING_RESPONSE = 4
+    while True:
+        state, status = advance_connection(pgconn)
+        if status == CONNECTION_AWAITING_RESPONSE:
+            nbytes = 8096
+            if libpq.PQsslInUse(pgconn.pgconn_ptr):
+                buffer = bytearray(nbytes)
+                c_buffer = create_string_buffer(bytes(buffer), nbytes)
+                ssl_obj = libpq.PQgetssl(pgconn.pgconn_ptr)
+                nread = libssl.SSL_read(ssl_obj, c_buffer, nbytes)
+                challenge = bytearray(c_buffer.raw[:nread])
+            else:
+                challenge = sock.recv(nbytes)
+            for index, byte in enumerate(challenge):
+                msg_size = read_int32_from_bytes(challenge, index + 1)
+                auth_type = read_int32_from_bytes(challenge, index + 5)
+                AUTH_MD5 = 5
+                if byte == ord("R") and msg_size == 12 and auth_type == AUTH_MD5:
+                    salt = challenge[index + 9 : index + 9 + 4]
+                    return salt
+        elif state == psycopg2.extensions.POLL_OK:
+            raise Exception("Connection already established")
+
+
+def advance_until_end(pgconn):
+    while True:
+        state, status = advance_connection(pgconn)
+        if state == psycopg2.extensions.POLL_OK:
+            return
+
+
+def send_hash(pgconn, hash):
+    sock = fromfd(pgconn.fileno(), keep_fd=True)
```

Comment:
curious -- in golang, we would prefer to keep the package name so we would call `socketfromfd.fromfd()`. This helps the reader understand where this function is coming from without going back to the imports at the top. Is the idea here that the file is short enough that its not a big deal, or is this the standard pythonic way to handle things?

Reply:
there is not necessarily a standard for this case. I like it because I can look at the imports and know immediately what functions exactly i am using from each module/external library. of course what you said is also a disadvantage of this approach.